src/http.cc

```diff
@@ -99,14 +99,22 @@ HttpStateData::HttpStateData(FwdState *theFwdState) :
     if (fwd->serverConnection() != NULL)
         _peer = cbdataReference(fwd->serverConnection()->getPeer());         /* might be NULL */
 
+    // XXX: FwdState::connectDone() uses request->flags.sslPeek instead
+    const auto https =
+        request->flags.sslBumped || // HTTP request inside a bumped connection
+        request->url.getScheme() == AnyP::PROTO_HTTPS; // GET https://...
+
+    flags.peering =  _peer;
+    flags.tunneling = (_peer && https);
```

Comment:
@yadij 
Yes there is bug with "GET https://" requests.

Unpatched squid when receives such requests forward them as is (send an "GET https://") to the upstream peer proxy. 
This patch tries to change this behaviour. Squid establish a CONNECT tunnel through upstream proxy and uses SecureConnectors to connect to the origin server. 

I suppose we should keep the current  behaviour for "GET https://" requests. Please fix me if I am not correct.

(Also see my next comment about serious bugs has this patch with client-first bumped requests )

Reply:
@rousskov 
Amos comment show me a serious bug we have with client first bumped requests:
  - Squid bumps using client first a CONNECT request. 
  - The client sends the first "GET https://origin1" through bumped connection
  - Squid uses a peer to forward the GET "https://origin1" request. Thus
      * Establish a connection to peer
       * sends a CONNECT to this connection
       * establish a TLS tunnel to origin server
  - when finishes stores the TLS established connection to the peer to pconn list !!!! 
  - A new "GET https://origin2" request is comming
  - Inside FwdState::connectStart retrieves the previous one TLS connection using the pconnPop
  - Send the request destined to origin2 to the origin1 server :-)

Two bugs here:
  - We must store to pconnlist the connections opened from Tunneler.
  - The client-first bumped connections must never try to reuse a pconn connection.

Reply:
### GET https

*If* doing so does not require a large development effort, then we should preserve the old forwarding behavior for `GET https`. That old behavior itself is correct or wrong, depending on the deployment environment (i.e., it is a proxy/admin choice), but this specific PR is _not_ about changing that old behavior or making it configurable, so _preserving_ it is the right thing to do when all the other factors are equal.

### step1 bumping and pconnPool

> when finishes stores the TLS established connection to the peer to pconn list

That sounds correct, provided it supplies the right storage key -- this secure HTTP connection is to `origin1:443` (or equivalent), not `peer1:3128` (or equivalent). HTTP connections should be pooled at HTTP, not TCP level and, in addition to their HTTP destination, should know whether they are encrypted. Do we support the right keys already, in some shape of form?

> We must store to pconnlist the connections opened from Tunneler.

It sounds like you meant to say "must not". If we can supply the right connection key, then I think we should store those connections (because they are perfectly reusable). If we cannot, then we clearly should not store them (because we cannot correctly find them later).

> The client-first bumped connections must never try to reuse a pconn connection.

If they can ask for the _right_ connection, then I disagree. If they cannot, then I agree.

Reply:
> That sounds correct, provided it supplies the right storage key -- this secure HTTP connection is to origin1:443 (or equivalent), not peer1:3128

We are storing using the peer1:3128 key. But I do not think we can do it differently. Storing using the origin1:443 key to a connection which is actually an origin1:443_through_peer1:3128 connection does not looks correct.
Because if a new request for the origin1:443 comes, the cache_peer_access acls may decide that we should connect through a different peer, the peer_2:3128.

Reply:
Yes, the connection pool has to be peer-aware (or we need peer-specific connection pools). For the former, the key is a combination of HTTP destination (e.g., origin1:443 in our case), peer (e.g., peer1:3128 in our case), and probably some kind of "secured" or "needs TLS" setting.

I assume Squid connection pools do not support that today. Adding such support is outside this PR scope. Please disable pooling of step1-bumped through-peer connections (both lookup in and addition to the pool).

Reply:
In the other hand this is has the bad effect that for every client-first bumped HTTP request, through a cache_peer a new connection is needed.
A solution of this maybe is to use pinned connections for client-first bumped requests at least when a cache_peer is used. The client-first bumping mode is the only bumping mode which does not use pinned connections.

Reply:
Since step1 bumping is or ought to be a rare exception, I do not recommend delaying/complicating this PR to optimize that mode of operation. We are not making anything worse by not optimizing that mode. That (complex) optimization may be implemented later if needed.

Reply:
"GET https" issue resolved at 646e43d9

The step1 (client-first) bumping issue resolved at facef8baf

Reply:
FYI: the "GET https://" which arrive over plain-text are fine being sent to plain-text peers, that is effectively neutral in terms of security. Upgrading to secure is better, but admin policy should (and can with cache_peer_access) decide on whether that is required. On the other hand sending any bumped TLS/SSL request directly to a plain-text peer *is* a bug, regardless of which type of bumping happened.

Please do not use pinning as a way to retain these connections. That will lead down a slippery slope of other workarounds to fix problems with the temporary behaviour. Having a new connection per-request is okay to start with, and matches what other clients do.

That said, it does not seem to be much to add a PconnPool member CachePeer::idleTunnels with entries indexed by their origin:443 domain. Both peer object and desired origin are available to FwdState::connectStart() where the pool lookup would need to be done.

Reply:
> FYI: the "GET https://" which arrive over plain-text are fine being sent to plain-text peers

The desired level of security when satisfying a `GET https` request via a cache_peer is a matter of local policy; it is not necessarily determined by the "cache_peer ssl" option. Forwarding "as is" (i.e. without encrypting and tunneling) is "fine" in some cases and "not fine" in others. Fortunately, this PR does not need to improve `GET https` support. Preserving the current status quo is sufficient.

> it does not seem to be much to add a PconnPool member CachePeer::idleTunnels

Yes, adding a data member is trivial. However, correctly managing a large number of persistent connection pools (in some configurations) tied to somewhat ephemeral (in some deployments) CachePeer objects is a serious change that should be done in a dedicated PR, with all the associated planning, justification, testing, etc.

@yadij, AFAICT, Christos believes that he has addressed this change request. If you consider this change request addressed, please resolve this conversation. If you think there is something specific (related to this change request) that still should be done, please say so explicitly.

Reply:
I'm not seeing anything obviously broken with the new code.